wasm/src/lib.rs

```diff
@@ -0,0 +1,148 @@
+extern crate ed25519_dalek;
+extern crate hex;
+
+mod utils;
+
+use wasm_bindgen::prelude::*;
+
+use std::convert::TryFrom;
+use ed25519_dalek::{Keypair, PublicKey, SecretKey, Signature, Signer as _, Verifier as _, SECRET_KEY_LENGTH, PUBLIC_KEY_LENGTH};
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+
+cfg_if::cfg_if! {
+	if #[cfg(consolelog)] {
+		use web_sys::console;
+
+		fn console_log(info: &str) {
+			console::log_1(&info.into());
+		}
+	} else {
+		fn console_log(_info: &str) {}
+	}
+}
+
+#[wasm_bindgen]
+pub fn setup() {
+	console_log("@solana/wasm Initialized. WASM ready for calls.");
+	utils::set_panic_hook();
+}
+
+#[wasm_bindgen]
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct Pair{
+	public_key: Vec<u8>,
+	secret_key: Vec<u8>
+}
+
+/// Generate key pair for use with ED25519 from ed25519_dalek
+///
+/// * returned struct has two fields public and secret.
+#[wasm_bindgen(js_name = generateKeyPair)]
+pub fn generate_keypair() -> Result<JsValue, JsValue> {
+	// using get random directly instead of OsRng since it's easier to integrate with WASM
+	let mut seed = [0u8; 32];
+	getrandom::getrandom(&mut seed).unwrap();
+	let secret = SecretKey::from_bytes(&seed).unwrap();
```

Comment:
unwrap will just bubble as exception to javascript with callstack

Reply:
Ah yeah, you're totally right. A panic in wasm will bubble up through the wasm-bindgen bindings. The panic does look quite scary though from the wasm-bindgen error logging so I think it would be great if you returned errors properly. `wasm-bindgen` methods return `Result` so that Result::Err variants can be turned into thrown exceptions.

```
  console.error
    panicked at 'called `Result::unwrap()` on an `Err` value: ()', src/lib.rs:49:25
    
    Stack:
    
    Error: 
        at /Users/jstarry/Workspace/solana/solana/wasm/dist/solana-wasm.cjs.js:515:15
        at /Users/jstarry/Workspace/solana/solana/wasm/dist/solana-wasm.cjs.js:370:22
        at console_error_panic_hook::Error::new::h8677d9788b6f7488 (wasm-function[1080]:0x768bd)
        at console_error_panic_hook::hook_impl::hd059b7ee5c601848 (wasm-function[202]:0x5847e)
        at console_error_panic_hook::hook::h2376c682cbcd7b97 (wasm-function[1210]:0x78614)
        at core::ops::function::Fn::call::hfc902b7dbd494ca7 (wasm-function[981]:0x75058)
        at std::panicking::rust_panic_with_hook::h3e6619a8809a443a (wasm-function[346]:0x629cb)
        at rust_begin_unwind (wasm-function[1099]:0x76d6e)
        at core::panicking::panic_fmt::ha8209066b105fcdd (wasm-function[1311]:0x79a37)
        at core::option::expect_none_failed::ha72d60de35a76889 (wasm-function[569]:0x6ba30)

      506 | const __wbg_error_4bb6c2a97407129a = logError(function(arg0, arg1) {
      507 |     try {
    > 508 |         console.error(getStringFromWasm0(arg0, arg1));
          |                 ^
      509 |     } finally {
      510 |         wasm.__wbindgen_free(arg0, arg1);
      511 |     }

      at dist/solana-wasm.cjs.js:508:17
      at dist/solana-wasm.cjs.js:370:22
      at console_error_panic_hook::error::hcd67c963013e0f67 (wasm-function[412]:0x65e91)
      at console_error_panic_hook::hook_impl::hd059b7ee5c601848 (wasm-function[202]:0x58567)
      at console_error_panic_hook::hook::h2376c682cbcd7b97 (wasm-function[1210]:0x78614)
      at core::ops::function::Fn::call::hfc902b7dbd494ca7 (wasm-function[981]:0x75058)
      at std::panicking::rust_panic_with_hook::h3e6619a8809a443a (wasm-function[346]:0x629cb)
```